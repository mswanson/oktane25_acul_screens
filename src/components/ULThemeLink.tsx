import * as React from "react";

import { Link, type LinkProps } from "@/components/ui/link";
import { cn } from "@/lib/utils";
import { extractTokenValue } from "@/utils/helpers/tokenUtils";

export interface ULThemeLinkProps extends LinkProps {
  /**
   * The content of the link element.
   */
  children: React.ReactNode;
  /**
   * Additional class names for custom styling.
   */
  className?: string;
  /**
   * Optional flag to disable the link.
   */
  disabled?: boolean;
  /**
   * Optional ref for the link element.
   */
  ref?: React.Ref<HTMLAnchorElement>;
}

const ULThemeLink = ({
  children,
  className,
  disabled = false,
  ref,
  ...props
}: ULThemeLinkProps) => {
  // Base component styles
  const baseStyles =
    "text-link-focus text-(length:--ul-theme-font-links-size) font-(weight:--ul-theme-font-links-weight) focus:rounded-(--ul-theme-border-links-border-radius) hover:text-link-focus/80";

  // Disabled state styles
  const disabledStyles = disabled
    ? "pointer-events-none text-muted cursor-not-allowed"
    : "";

  // UL theme overrides
  const variantThemeOverrides =
    "theme-universal:focus:outline-none theme-universal:focus:ring-4 theme-universal:focus:ring-base-focus/15 theme-universal:focus:bg-base-focus/15"; // focus base color

  // Using extractTokenValue utility to extract the link style variant type from the CSS variable
  const linkStyleValue =
    extractTokenValue("--ul-theme-font-links-style") === "normal"
      ? "none"
      : "always";

  return (
    <Link
      ref={ref}
      className={cn(
        baseStyles,
        disabledStyles,
        variantThemeOverrides,
        className
      )}
      underline={linkStyleValue}
      aria-disabled={disabled}
      {...props}
    >
      {children}
    </Link>
  );
};

export default ULThemeLink;
