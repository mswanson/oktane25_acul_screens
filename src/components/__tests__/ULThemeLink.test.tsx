import { fireEvent, render, screen } from "@testing-library/react";

import ULThemeLink from "../ULThemeLink";

describe("ULThemeLink Component", () => {
  const defaultProps = {
    href: "https://example.com",
    children: "Example Link",
    className: "custom-class",
  };

  it("renders correctly and matches snapshot", () => {
    const { container } = render(<ULThemeLink {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  it("renders the correct text", () => {
    render(<ULThemeLink {...defaultProps} />);
    expect(screen.getByText("Example Link")).toBeInTheDocument();
  });

  it("applies additional class names", () => {
    render(<ULThemeLink {...defaultProps} />);
    const link = screen.getByRole("link");
    expect(link).toHaveClass("custom-class");
  });

  it("renders the correct href attribute", () => {
    render(<ULThemeLink {...defaultProps} />);
    const link = screen.getByRole("link");
    expect(link).toHaveAttribute("href", "https://example.com");
  });

  it("calls the onClick handler when clicked", () => {
    const mockOnClick = jest.fn();
    render(<ULThemeLink {...defaultProps} onClick={mockOnClick} />);
    const link = screen.getByRole("link");
    fireEvent.click(link);
    expect(mockOnClick).toHaveBeenCalledTimes(1);
  });

  it("disables the link when the disabled prop is true", () => {
    render(<ULThemeLink {...defaultProps} disabled />);
    const link = screen.getByRole("link");
    expect(link).toHaveClass("pointer-events-none cursor-not-allowed");
    expect(link).toHaveAttribute("aria-disabled", "true");
  });
});
