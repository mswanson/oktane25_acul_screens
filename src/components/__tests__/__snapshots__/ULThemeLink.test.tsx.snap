// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ULThemeLink Component renders correctly and matches snapshot 1`] = `
<div>
  <a
    aria-disabled="false"
    class="focus-within:ring-ring inline-flex items-center gap-2 rounded-md py-0.5 underline-offset-4 transition-colors focus:ring-3 focus-visible:outline-hidden underline text-link-focus text-(length:--ul-theme-font-links-size) font-(weight:--ul-theme-font-links-weight) focus:rounded-(--ul-theme-border-links-border-radius) hover:text-link-focus/80 theme-universal:focus:outline-none theme-universal:focus:ring-4 theme-universal:focus:ring-base-focus/15 theme-universal:focus:bg-base-focus/15 custom-class"
    href="https://example.com"
  >
    Example Link
  </a>
</div>
`;
